File: java/src/com/nttdocomo/util/Timer.java
package com.nttdocomo.util;

import java.util.TimerTask;

public class Timer {

    private TimerListener listener;
    private int delayMillis;
    private boolean repeat;

    private java.util.Timer internalTimer;
    private TimerTask task;
    private boolean running;

    public Timer() {
        this.running = false;
    }

    public void setTime(int millis) {
        this.delayMillis = millis;
    }

    public void setRepeat(boolean repeat) {
        this.repeat = repeat;
    }

    public void setListener(TimerListener listener) {
        this.listener = listener;
    }

    public synchronized void start() {
        if (running) {
            throw new IllegalStateException("Timer already running");
        }

        if (listener == null) {
            throw new IllegalStateException("TimerListener not set");
        }

        internalTimer = new java.util.Timer();
        running = true;

        task = new TimerTask() {
            @Override
            public void run() {
                listener.timerExpired(Timer.this);

                if (!repeat) {
                    stop();
                }
            }
        };

        if (repeat) {
            internalTimer.schedule(task, delayMillis, delayMillis);
        } else {
            internalTimer.schedule(task, delayMillis);
        }
    }

    public synchronized void stop() {
        running = false;

        if (task != null) {
            task.cancel();
            task = null;
        }

        if (internalTimer != null) {
            internalTimer.cancel();
            internalTimer = null;
        }
    }
}

File: web/src/com/nttdocomo/util/Timer.ts
import type { TimerListener } from './TimerListener';

export class Timer {
	private listener: TimerListener | null = null;
	private delayMillis = 0;
	private repeat = false;

	private handle:
		| ReturnType<typeof setTimeout>
		| ReturnType<typeof setInterval>
		| null = null;

	private running = false;

	public setTime(millis: number): void {
		this.delayMillis = millis;
	}

	public setRepeat(repeat: boolean): void {
		this.repeat = repeat;
	}

	public setListener(listener: TimerListener): void {
		this.listener = listener;
	}

	public start(): void {
		if (this.running) {
			throw new Error('Timer already running');
		}
		if (!this.listener) {
			throw new Error('TimerListener not set');
		}

		this.running = true;

		if (this.repeat) {
			this.handle = setInterval(() => {
				this.listener?.timerExpired(this);
			}, this.delayMillis);
		} else {
			this.handle = setTimeout(() => {
				this.listener?.timerExpired(this);
				this.stop();
			}, this.delayMillis);
		}
	}

	public stop(): void {
		this.running = false;

		if (this.handle != null) {
			if (this.repeat) {
				clearInterval(this.handle as ReturnType<typeof setInterval>);
			} else {
				clearTimeout(this.handle as ReturnType<typeof setTimeout>);
			}
			this.handle = null;
		}
	}
}
